### PR TITLE
Let phpcsfixer2 select the correct files and improve performance

### PR DESCRIPTION
**Pull request by veewee**
| Q             | A
| ------------- | ---
| Branch        | master for features and deprecations
| Bug fix?      | yes/no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Documented?   | yes
| Fixed tickets | #371, #327

### CODE REVIEW

**Comment by veewee** [comment]
Hi @Landerstraeten , @keradus,

This is an attempt to fix the issues in #371 and #327.
Can you give it a review to make sure this works as expected?

Thanks!

File: spec/Task/PhpCsFixerV2Spec.php [hunk 1]
```diff
@@ -107,35 +106,8 @@ function it_runs_the_suite_for_all_files(
         $result->isPassed()->shouldBe(true);
     }
 
-    function it_runs_the_suite_for_changed_files(
```
Comment (keradus) [inline spec/Task/PhpCsFixerV2Spec.php]: why dropping the test instead of replacing it with test of new feature ?

File: doc/tasks/phpcsfixer2.md [hunk 2]
```diff
@@ -97,12 +97,13 @@ The tool will fix all files if the tool version has changed or the list of fixer
...
+**can_intersect**
 
-*Default: null*
-
-Specify path mode (can be override or intersection).
+*Default: true*
 
+Intersection mode can only be used when you have a configuration file which contains a Finder.
+This mode works best since only files that are being commit and are in your configuration will be checked.
+When there is no Finder in your configuration, all committed files will be checked.
```
Comment (keradus) [inline doc/tasks/phpcsfixer2.md]: when there is no finder in configuration, setting path-mode=intersection will crash the execution

File: src/Task/PhpCsFixerV2.php [hunk 3]
```diff
@@ -82,15 +89,23 @@ public function run(ContextInterface $context)
         }
 
         $arguments->addOptionalArgument('--using-cache=%s', $config['using_cache'] ? 'yes' : 'no');
-        $arguments->addOptionalArgument('--path-mode=%s', $config['path_mode']);
+        $arguments->addOptionalArgument('--path-mode=intersect', $config['can_intersect']);
```
Comment (keradus) [inline src/Task/PhpCsFixerV2.php]: did you try this out? please add tests to cover this change, as it is not working.
```diff
-intersect
+intersection
```

Comment (veewee) [inline src/Task/PhpCsFixerV2.php]: We currently don't have a system to do e2e tests with 3rd party tools. I looked over that one during commit and will test these changes thoroughly before merging.

**Comment by veewee** [comment]
Thanks for your feedback @keradus!

I've changed the logic a bit:

- Run context with can_intersect=true will never receive the files since the finder in the php-cs-fixer2 config can be used. (Works better on windows since they have a limited CLI input charakters)
- Run context with can_intersect=false will always retrieve the files since there is no finder in configuration
- PreCommit context with can_intersect=true will always retrieve the files and will run in intersection path mode. When there is no finder in the config file, the exectuable will fail.
- PreCommit context with can_intersect=false will always retrieve the files and will run in override path mode

I assume this will cover all cases at the moment, but will test this thoroughly.

File: doc/tasks/phpcsfixer2.md [hunk 4]
```diff
@@ -103,7 +103,8 @@ Cache is supported only for tool downloaded as phar file or installed via compos
 
 Intersection mode can only be used when you have a configuration file which contains a Finder.
 This mode works best since only files that are being commit and are in your configuration will be checked.
-When there is no Finder in your configuration, all committed files will be checked.
+When there is no Finder in your configuration, you'll have set this parameter to false. 
+Otherwise php-cs-fixer will crash the execution.
```
Comment (keradus) [inline doc/tasks/phpcsfixer2.md]: btw, could there be consistency with naming? sometimes you use `PHP-CS-Fixer`, sometimes you use `php-cs-fixer` after binary file, while official name is `PHP Coding Standards Fixer` or `PHP CS Fixer` when one used shorter version.

File: src/Task/PhpCsFixerV2.php [hunk 5]
```diff
@@ -89,11 +89,14 @@ public function run(ContextInterface $context)
...
         $arguments->addOptionalArgument('--using-cache=%s', $config['using_cache'] ? 'yes' : 'no');
-        $arguments->addOptionalArgument('--path-mode=intersect', $config['can_intersect']);
+        $arguments->addOptionalArgument('--path-mode=intersection', $config['can_intersect']);
         $arguments->addOptionalArgument('--verbose', $config['verbose']);
         $arguments->addOptionalArgument('--diff', $config['diff']);
         $arguments->add('fix');
-        $arguments->addFiles($files);
+
+        if ($context instanceof GitPreCommitContext || !$config['can_intersect']) {
+            $arguments->addFiles($files);
```
Comment (keradus) [inline src/Task/PhpCsFixerV2.php]: not sure about condition here. why files are not used if `can_intersect=true` ?

**Comment by keradus** [comment]
> - Run context with can_intersect=true will never receive the files since the finder in the php-cs-fixer2 config can be used. (Works better on windows since they have a limited CLI input charakters)
> - Run context with can_intersect=false will always retrieve the files since there is no finder in configuration

Then, what is the purpose of passing files in grumphp integration together with can_intersect=true ?

you are trying to use some extra logic on top of integration changing decision of user.
if user passed files, let use it.
if user configure mode=intersection, let use it.
that's it

File: doc/tasks/phpcsfixer2.md (hunk 4 above)
Comment (veewee) [inline doc/tasks/phpcsfixer2.md]: Sure, I'll try to keep that in mind.

**Comment by veewee** [comment]
> Then, what is the purpose of passing files in grumphp integration together with can_intersect=true ?
> you are trying to use some extra logic on top of integration changing decision of user.
> if user passed files, let use it.
> if user configure mode=intersection, let use it.
> that's it


I disagree: 

In run context, all files are being checked. That means that the task contains ALL files known to git. It is a bit overkill to paste them all the the CLI when there is a finder config in PHP CS Fixer. Besides that: Windows can't handle too many input characters.
So as long as there is a finder in configuration (can_intersect=true), we shouldn't pass the files in run context in my opninion.

The PreCommit context only contains the changed files. In this context, the intersection path-mode is the best option. If this is not possible (can_intersect=false), the override path-mode will be used.

Does this makes sense to you or would you still drop that logic?

**Comment by keradus** [comment]
> So as long as there is a finder in configuration (can_intersect=true), we shouldn't pass the files in run context in my opninion.

so, why this scenario is allowed by code ?

**Comment by veewee** [comment]
I don't really understand your question. Can you give me some additional context?

**Comment by keradus** [comment]
```php
    if ($context instanceof GitPreCommitContext || !$config['can_intersect']) {
        $arguments->addFiles($files);
    }
```

let discuss not-precommit context, thus we have:
```php
    if (!$config['can_intersect']) {
        $arguments->addFiles($files);
    }
```

vs

> In run context, all files are being checked.

No, looking at code, in run-context, the paths are sometimes passed, sometimes not

**Comment by keradus** [comment]
could `can_intersect` be different for run-context and precommit-context?

**Comment by veewee** [comment]
@keradus, I see what you are trying to say. Probably I explained incorrect:

`can_intersect` can't be different for run or pre-commit context, it is always the same since it is in configration. Maybe a better name would be `config_contains_finder` or something like that.

So lets asume `can_intersect` will be renamed to `config_contains_finder`. In that case you have this if:

```php
    if ($context instanceof GitPreCommitContext || !$config['config_contains_finder']) {
        $arguments->addFiles($files);
    }
```

Let's focus on RunContext:
```
    if (!$config['config_contains_finder']) {
        $arguments->addFiles($files);
    }
```

As you can see: the files are being added when there is no finder configuration. If we don't do this, PHP CS Fixer will crash since it isn't validating any files.
When there is a finder in configuration, this one can be used instead of adding all files. This is because it should run on all files configured in the finder anyway.


Off course: for PreComit context, we'll always pass the files since we only want to check a limited subset.

**Comment by keradus** [comment]
cleaner to me, yet I do still not clear enough.

> RunContext:
> As you can see: the files are being added when there is no finder configuration. If we don't do this, PHP CS Fixer will crash since it isn't validating any files.

great, I do agree that if there is no finder, we shall pass the files.

then, you have execution scenario:
- context: RunContext
- config_contains_finder: true
what is gonna happen with current implementation?
you gonna configure `--path-mode=intersection` yet no files for it, and intersecting finder with empty set is... empty set.

File: doc/tasks/phpcsfixer2.md [hunk 6]
```diff
@@ -97,12 +97,14 @@ The tool will fix all files if the tool version has changed or the list of fixer
 Cache is supported only for tool downloaded as phar file or installed via composer.
 
 
-**path_mode**
+**can_intersect**
```
Comment (Landerstraeten) [inline doc/tasks/phpcsfixer2.md]: I think you forgot to change the example on line 25

File: resources/config/tasks.yml [hunk 7]
```diff
@@ -232,7 +232,6 @@ services:
         arguments:
           - '@config'
           - '@process_builder'
-          - '@async_process_runner'
```
Comment (Landerstraeten) [inline resources/config/tasks.yml]: You deleted the argument from another task. Delete line 244 instead.

File: doc/tasks/phpcsfixer2.md [hunk 8]
```diff
@@ -97,12 +97,14 @@ The tool will fix all files if the tool version has changed or the list of fixer
...
 
-*Default: null*
-
-Specify path mode (can be override or intersection).
+*Default: true*
 
+Intersection mode can only be used when you have a configuration file which contains a Finder.
+This mode works best since only files that are being commit and are in your configuration will be checked.
+When there is no Finder in your configuration, you'll have set this parameter to false. 
+Otherwise php-cs-fixer will crash the execution.
```
Comment (Landerstraeten) [inline doc/tasks/phpcsfixer2.md]: This is clear 👍

**Comment by veewee** [comment]
Fixed in #395

**Pull request closed without merging**